### ISSUE TITLE
[BUG]  NaN Loss Error in NNFastAiTabularModel When Continuous Feature Has Zero Variance and Missing Values

### ISSUE THREAD

**Issue by amirbalef** [autogluon/autogluon#4508]
Hello,
I encountered an issue when using the NNFastAiTabularModel. Specifically, when dealing with a continuous feature that has missing values and zero variance, I receive the following error during training:

```python
AssertionError: Exception occured in `EarlyStoppingCallbackWithTimeLimit` when calling event `after_epoch`:
	WARNING: NaN loss encountered in epoch 0!
```

**Expected behavior**

I believe the issue arises in line 180 of the following file:
[autogluon/tabular/models/fastainn/tabular_nn_fastai.py#L182](https://github.com/autogluon/autogluon/blob/e0b111cba4b797456fdcaf12750af56e80de1313/tabular/src/autogluon/tabular/models/fastainn/tabular_nn_fastai.py#L182)

A temporary fix that worked for me was adding this to before the line :

```python
cont_std[cont_std==0] = 1e-10
```

**To Reproduce**
Here is a minimal example to reproduce the issue:
```python
from autogluon.tabular.models import NNFastAiTabularModel
import numpy as np
import pandas as pd

def generate_xor_data(n_samples=100, feature_num=3):
    X = np.random.rand(n_samples, feature_num)
    X[:, 1] = 0.5
    X[0, 1] = np.nan
    df = pd.DataFrame(X, columns=["input_" + str(i) for i in range(feature_num)])
    df["class"] = np.logical_xor(X[:, 0] > 0.5, X[:, 2] > 0.5).astype(int)
    return df

df = generate_xor_data(n_samples=300)

model = NNFastAiTabularModel(
    path="./autogluon_output", problem_type="binary", eval_metric="roc_auc"
)
model.fit(
    X=df.drop(columns=["class"]),
    y=df["class"],
    num_cpus=1,
    num_gpus=0,
    verbosity=4,
)
```

**Installed Versions**
<details>
INSTALLED VERSIONS
------------------
date                   : 2024-10-02
time                   : 08:47:52.932468
python                 : 3.10.15.final.0
OS                     : Linux
OS-release             : 6.5.0-1027-oem
Version                : #28-Ubuntu SMP PREEMPT_DYNAMIC Thu Jul 25 13:32:46 UTC 2024
machine                : x86_64
processor              : x86_64
num_cores              : 20
cpu_ram_mb             : 63940.265625
cuda version           : None
num_gpus               : 0
gpu_ram_mb             : []
avail_disk_size_mb     : 350069

accelerate             : 0.21.0
autogluon              : 1.1.1
autogluon.common       : 1.1.1
autogluon.core         : 1.1.1
autogluon.features     : 1.1.1
autogluon.multimodal   : 1.1.1
autogluon.tabular      : 1.1.1
autogluon.timeseries   : 1.1.1
boto3                  : 1.35.30
catboost               : 1.2.7
defusedxml             : 0.7.1
evaluate               : 0.4.1
fastai                 : 2.7.17
gluonts                : 0.14.3
hyperopt               : 0.2.7
imodels                : None
jinja2                 : 3.1.4
joblib                 : 1.4.2
jsonschema             : 4.23.0
lightgbm               : 4.3.0
lightning              : 2.3.3
matplotlib             : 3.9.2
mlforecast             : 0.10.0
networkx               : 3.3
nlpaug                 : 1.1.11
nltk                   : 3.9.1
nptyping               : 2.4.1
numpy                  : 1.26.4
nvidia-ml-py3          : None
omegaconf              : 2.3.0
onnxruntime-gpu        : None
openmim                : 0.3.7
optimum                : None
optimum-intel          : None
orjson                 : 3.10.7
pandas                 : 2.2.3
pdf2image              : 1.17.0
Pillow                 : 10.4.0
psutil                 : 5.9.8
pytesseract            : 0.3.10
pytorch-lightning      : 2.3.3
pytorch-metric-learning: 2.3.0
ray                    : 2.31.0
requests               : 2.32.3
scikit-image           : 0.20.0
scikit-learn           : 1.5.2
scikit-learn-intelex   : None
scipy                  : 1.14.1
seqeval                : 1.2.2
setuptools             : 75.1.0
skl2onnx               : None
statsforecast          : 1.4.0
tabpfn                 : None
tensorboard            : 2.18.0
text-unidecode         : 1.3
timm                   : 0.9.16
torch                  : 2.4.0.post101
torchmetrics           : 1.2.1
torchvision            : 0.19.1a0+ee30128
tqdm                   : 4.66.5
transformers           : 4.38.2
utilsforecast          : 0.0.10
vowpalwabbit           : None
xgboost                : 2.1.1
</details>

**Comment by rsj123** [comment]
I met similar issue in NeuralNetFastAI

> /root/miniconda3/lib/python3.10/site-packages/autogluon/tabular/models/fastainn/tabular_nn_fastai.py:182: RuntimeWarning: invalid value encountered in divide
>   (X[self.cont_columns].values - cont_mean) / cont_std,
> /root/miniconda3/lib/python3.10/site-packages/autogluon/tabular/models/fastainn/tabular_nn_fastai.py:182: RuntimeWarning: divide by zero encountered in divide
>   (X[self.cont_columns].values - cont_mean) / cont_std,
> /root/miniconda3/lib/python3.10/site-packages/autogluon/tabular/models/fastainn/tabular_nn_fastai.py:182: RuntimeWarning: invalid value encountered in divide
>   (X[self.cont_columns].values - cont_mean) / cont_std,
>         Warning: Exception caused NeuralNetFastAI to fail during training... Skipping this model.
>                 Exception occured in `EarlyStoppingCallbackWithTimeLimit` when calling event `after_epoch`:
>         WARNING: NaN loss encountered in epoch 0!
> Detailed Traceback:
> Traceback (most recent call last):
>   File "/root/miniconda3/lib/python3.10/site-packages/autogluon/core/trainer/abstract_trainer.py", line 1904, in _train_and_save
>     model = self._train_single(X, y, model, X_val, y_val, total_resources=total_resources, **model_fit_kwargs)
>   File "/root/miniconda3/lib/python3.10/site-packages/autogluon/core/trainer/abstract_trainer.py", line 1844, in _train_single
>     model = model.fit(X=X, y=y, X_val=X_val, y_val=y_val, total_resources=total_resources, **model_fit_kwargs)
>   File "/root/miniconda3/lib/python3.10/site-packages/autogluon/core/models/abstract/abstract_model.py", line 856, in fit
>     out = self._fit(**kwargs)
>   File "/root/miniconda3/lib/python3.10/site-packages/autogluon/tabular/models/fastainn/tabular_nn_fastai.py", line 359, in _fit
>     self.model.fit_one_cycle(epochs, params["lr"], cbs=callbacks)
>   File "/root/miniconda3/lib/python3.10/site-packages/fastai/callback/schedule.py", line 121, in fit_one_cycle
>     self.fit(n_epoch, cbs=ParamScheduler(scheds)+L(cbs), reset_opt=reset_opt, wd=wd, start_epoch=start_epoch)
>   File "/root/miniconda3/lib/python3.10/site-packages/fastai/learner.py", line 266, in fit
>     self._with_events(self._do_fit, 'fit', CancelFitException, self._end_cleanup)
>   File "/root/miniconda3/lib/python3.10/site-packages/fastai/learner.py", line 201, in _with_events
>     try: self(f'before_{event_type}');  f()
>   File "/root/miniconda3/lib/python3.10/site-packages/fastai/learner.py", line 255, in _do_fit
>     self._with_events(self._do_epoch, 'epoch', CancelEpochException)
>   File "/root/miniconda3/lib/python3.10/site-packages/fastai/learner.py", line 203, in _with_events
>     self(f'after_{event_type}');  final()
>   File "/root/miniconda3/lib/python3.10/site-packages/fastai/learner.py", line 174, in __call__
>     def __call__(self, event_name): L(event_name).map(self._call_one)
>   File "/root/miniconda3/lib/python3.10/site-packages/fastcore/foundation.py", line 159, in map
>     def map(self, f, *args, **kwargs): return self._new(map_ex(self, f, *args, gen=False, **kwargs))
>   File "/root/miniconda3/lib/python3.10/site-packages/fastcore/basics.py", line 910, in map_ex
>     return list(res)
>   File "/root/miniconda3/lib/python3.10/site-packages/fastcore/basics.py", line 895, in __call__
>     return self.func(*fargs, **kwargs)
>   File "/root/miniconda3/lib/python3.10/site-packages/fastai/learner.py", line 178, in _call_one
>     for cb in self.cbs.sorted('order'): cb(event_name)
>   File "/root/miniconda3/lib/python3.10/site-packages/fastai/callback/core.py", line 64, in __call__
>     except Exception as e: raise modify_exception(e, f'Exception occured in `{self.__class__.__name__}` when calling event `{event_name}`:\n\t{e.args[0]}', replace=True)
>   File "/root/miniconda3/lib/python3.10/site-packages/fastai/callback/core.py", line 62, in __call__
>     try: res = getcallable(self, event_name)()
>   File "/root/miniconda3/lib/python3.10/site-packages/autogluon/tabular/models/fastainn/callbacks.py", line 63, in after_epoch
>     raise AssertionError(f"WARNING: NaN loss encountered in epoch {self.epoch}!")
> AssertionError: Exception occured in `EarlyStoppingCallbackWithTimeLimit` when calling event `after_epoch`:
>         WARNING: NaN loss encountered in epoch 0!

**Comment by Innixma** [comment]
Hi @amirbalef and @rsj123, thanks for highlighting this issue! I have fixed this bug in mainline and the fix will be included in the v1.2 release. Cheers!
